datafusion/sqllogictest/test_files/unnest.slt

```diff
@@ -383,5 +383,23 @@ select unnest(array_remove(column1, 3)) - 1 as c1, column3 from unnest_table;
 5 3
 11 NULL
 
+# Unnest with non-nullable lists
+statement ok
+create table non_nullable_list(
+    c1 int[] not null,
+    c2 int[] not null
+) as values ([1,2,3], [4]);
+
+# Unnesting may produce NULLs even if the lists are non-nullable
```

Comment:
Added the test suggested by @jayzhan211 in https://github.com/apache/arrow-datafusion/pull/10044#discussion_r1563882878